### PR TITLE
AutoGluon Modularization - separating core, tabular, vision, text into namespaces

### PR DESCRIPTION
**Pull request by gradientsky**
Please add your comments/suggestions for proposed layout. If you want to try this version:

Local install:
```shell
deactivate
rm -rf ./venv
python3 -m venv ./venv
source ./venv/bin/activate
pip install --upgrade pip
pip install mxnet-cu102
pip install gluonnlp==0.8.3
pip install torch
pip install torchvision
pip install wheel


# Mac-only fix
if [[ "$OSTYPE" == "darwin"* ]]; then
  pip uninstall typing -y
fi

./full_install.sh
```

Running tests:
```shell
cd ~/Projects/autogluon/core/
pytest
cd ~/Projects/autogluon/tabular/
pytest
cd ~/Projects/autogluon/mxnet/
pytest
cd ~/Projects/autogluon/extra/
pytest
cd ~/Projects/autogluon/text/
pytest
cd ~/Projects/autogluon/vision/
pytest
```

Running docs:
```shell
cd ~/Projects/autogluon/
python3 -m pip install git+https://github.com/zhanghang1989/d2l-book
python3 -m pip install --force-reinstall ipython==7.16
pip install jupyter_sphinx
pip install docutils\<0.16
cd ~/Projects/autogluon/docs && bash build_doc.sh
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -43,7 +82,43 @@ stage("Build Docs") {
         git clean -fx
         python3 -m pip install git+https://github.com/zhanghang1989/d2l-book
         python3 -m pip install --force-reinstall ipython==7.16
-        python3 -m pip install --upgrade --force-reinstall -e .
+
+       pip uninstall -y autogluon
```
Comment (sxjscience) [inline Jenkinsfile:86]: nit: seems to miss one space.

File: core/tests/unittests/bayesopt/gpmxnet/test_comparison_gpy.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
+# import os
```
Comment (sxjscience) [inline core/tests/unittests/bayesopt/gpmxnet/test_comparison_gpy.py:1]: Should we remove this file?

File: tabular/src/autogluon/tabular/utils/ml/trainer/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from ....utils.ml.trainer.auto_trainer import *
```
Comment (sxjscience) [inline tabular/src/autogluon/tabular/utils/ml/trainer/__init__.py:1]: It seems that we may need an empty line.

File: core/tests/unittests/bayesopt/gpmxnet/test_comparison_gpy.py (hunk 2 above)
Comment (mseeger) [reply core/tests/unittests/bayesopt/gpmxnet/test_comparison_gpy.py:1]: You can remove it, but I will then add it back in once the modularization is done. #695

**Comment by szha** [comment]
Job PR-694-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/4/index.html

File: core/tests/unittests/bayesopt/gpmxnet/test_comparison_gpy.py (hunk 2 above)
Comment (gradientsky) [reply core/tests/unittests/bayesopt/gpmxnet/test_comparison_gpy.py:1]: Removed

**Comment by jwmueller** [comment]
So the virtualenv and the autogluon folder need to be in the same place?
When not, I got this error during `pip install -e .`:

```
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode
```

**Comment by jwmueller** [comment]
Why do numpy and wheel need to be installed separately?

**Comment by gradientsky** [comment]
@jwmueller venv can be anywhere. I have everything under `~/Projects/` and autogluon is checked out in `~/Projects/autogluon/`. Each individual module has to be installed from sub-directory - see the script in PR description.

**Comment by szha** [comment]
Job PR-694-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/5/index.html

**Comment by szha** [comment]
Job PR-694-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/6/index.html

File: full_install.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
+deactivate
+rm -rf ./venv
+python3 -m venv ./venv
+source ./venv/bin/activate
+pip install -U pip
```
Comment (Innixma) [inline full_install.sh:5]: --upgrade

File: full_install.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+deactivate
+rm -rf ./venv
+python3 -m venv ./venv
```
Comment (Innixma) [inline full_install.sh:3]: why make a new venv? I would expect it to install on whatever environment is currently in use by the user.

File: full_install.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+source ./venv/bin/activate
+pip install -U pip
+pip install mxnet-cu102
+pip install numpy
+pip install gluonnlp==0.8.3
+pip install torch
```
Comment (Innixma) [inline full_install.sh:9]: Why install torch, torchvision, gluonnlp, numpy, wheel? Why uninstall typing?

File: full_install.sh [hunk 7]
```diff
@@ -0,0 +1,26 @@
+deactivate
+rm -rf ./venv
+python3 -m venv ./venv
+source ./venv/bin/activate
+pip install -U pip
+pip install mxnet-cu102
```
Comment (Innixma) [inline full_install.sh:6]: why install gpu mxnet? Why not just assume mxnet is installed prior to calling this script?

File: full_install.sh (hunk 4 above)
Comment (Innixma) [reply full_install.sh:5]: + We probably don't want this here, it should be in instructions that user should have already upgraded pip prior to calling this.

File: full_install.sh (hunk 7 above)
Comment (gradientsky) [reply full_install.sh:6]: Text/cv tests won't pass without gpu

File: full_install.sh (hunk 6 above)
Comment (gradientsky) [reply full_install.sh:9]: Typing is mac-only issue: it conflicts with python local installation. Other dependencies are needed for tests.

File: full_install.sh (hunk 5 above)
Comment (gradientsky) [reply full_install.sh:3]: We can remove this. I was primarily using this script for development.

File: full_install.sh (hunk 4 above)
Comment (gradientsky) [reply full_install.sh:5]: -U is the same as --upgrade

**Comment by szha** [comment]
Job PR-694-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/7/index.html

File: .github/workflows/pypi_release.yml [hunk 8]
```diff
@@ -30,5 +30,31 @@ jobs:
...
+        python setup.py sdist bdist_wheel
+        twine upload dist/* --verbose
+        cd ..
+        cd vision/
+        python setup.py sdist bdist_wheel
+        twine upload dist/* --verbose
+        cd ..
+        cd autogluon/
+        python setup.py sdist bdist_wheel
+        twine upload dist/* --verbose
```
Comment (jwmueller) [inline .github/workflows/pypi_release.yml:59]: Can you briefly spell out what'll happen when user executes `pip install autogluon`? (once everything's on pypi)

File: Jenkinsfile [hunk 9]
```diff
@@ -43,7 +82,43 @@ stage("Build Docs") {
...
+
+        cd mxnet/
+        python3 -m pip install --upgrade -e .
+        cd ..
+
+        cd extra/
+        python3 -m pip install --upgrade -e .
+        cd ..
+
+        cd text/
```
Comment (jwmueller) [inline Jenkinsfile:110]: Suppose a developer accidentally puts something in `text` that depends on dependency of `tabular` without adding dependency in `text`. How will this be detected, or is it something we just need to catch via code review?

File: README.md [hunk 10]
```diff
@@ -20,7 +20,7 @@ AutoGluon automates machine learning tasks enabling you to easily achieve strong
 # python3 -m pip install --upgrade "mxnet<2.0.0"
 # python3 -m pip install autogluon
```
Comment (jwmueller) [inline README.md:21]: will these still suffice for the example to run?

Comment (jwmueller) [reply README.md:21]: should we somewhere add explicit announcement that this major change in APIs from v0.14?

File: autogluon/setup.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        f.write("__version__ = '{}'\n".format(version))
+
+
+long_description = open(os.path.join('..', 'README.md')).read()
+
+MIN_PYTHON_VERSION = '>=3.6.*'
+
+requirements = [
+    'Pillow<=6.2.1',
+    'numpy>=1.16.0',
```
Comment (jwmueller) [inline autogluon/setup.py:44]: here there is min numpy version, but in the other install-script there is not

File: core/setup.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        f.write("__version__ = '{}'\n".format(version))
+
+
+long_description = open(os.path.join('..', 'README.md')).read()
+
+MIN_PYTHON_VERSION = '>=3.6.*'
+
+requirements = [
+    'Pillow<=6.2.1',
+    'numpy>=1.16.0',
```
Comment (jwmueller) [inline core/setup.py:44]: any idea why there are other numpy installs still needed beside this one?

File: core/src/autogluon/core/loss.py [hunk 13]
```diff
@@ -1,5 +1,5 @@
 from mxnet.gluon import loss
-from ..core import obj
+from . import obj
 
 __all__ = ['SoftmaxCrossEntropyLoss']
```
Comment (jwmueller) [inline core/src/autogluon/core/loss.py:5]: Seems like this file could be easily moved out of core to make it easier for core to become free of mxnet-dependency

**Comment by jwmueller** [comment]
So core will depend on mxnet for the foreseeable future?
Major mxnet items in there are the: bayesopt searcher (should be mxnet-dependency-free soon), RL-searcher (will not be mxnet free in the foreseeable future), miscellaneous mxnet items like losses (should maybe be moved to `mxnet/` instead of `core/` now already since it seems easy to do?)

File: core/src/autogluon/core/utils/utils.py [hunk 14]
```diff
@@ -5,26 +5,12 @@
...
 
 import numpy as np
-import pandas as pd
-from pandas import DataFrame, Series
-from sklearn.model_selection import KFold, StratifiedKFold, RepeatedKFold, RepeatedStratifiedKFold, train_test_split
-
-from .constants import BINARY, REGRESSION, MULTICLASS, SOFTCLASS
-from ..metrics import accuracy, root_mean_squared_error, Scorer
+from pandas import DataFrame
+from sklearn.model_selection import KFold, StratifiedKFold, RepeatedKFold, RepeatedStratifiedKFold
```
Comment (jwmueller) [inline core/src/autogluon/core/utils/utils.py:9]: Is `sklearn` often used elsewhere in core? Otherwise could consider moving these functions to reduce one more dependency

File: docs/api/index.rst [hunk 15]
```diff
@@ -8,13 +8,13 @@ describing what they are and what they do. The APIs are grouped into the followi
 
    .. card::
       :title: Prediction Tasks
-      :link: autogluon.task.html
+      :link: autogluon.core.task.html
```
Comment (jwmueller) [inline docs/api/index.rst:11]: documentation issue

File: Jenkinsfile [hunk 16]
```diff
@@ -16,9 +16,48 @@ stage("Unit Test") {
         export LD_LIBRARY_PATH=/usr/local/cuda-10.1/lib64
         export MPLBACKEND=Agg
         export MXNET_CUDNN_AUTOTUNE_DEFAULT=0
-        python3 -m pip install --upgrade --force-reinstall -e .
-        python3 -m pip install pytest
+
```
Comment (mli) [inline Jenkinsfile:19]: we can just remove the conda env, and rebuilt it for each test. one issue here is assigning a uniqle name to each executor to allow parallel testing, you could refer to d2l's setup https://github.com/d2l-ai/d2l-en/blob/master/Jenkinsfile#L14

File: core/setup.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    'tornado>=5.0.1',
+    'requests',
+    'matplotlib',
+    'tqdm>=4.38.0',
+    'paramiko>=2.4',
+    'dask>=2.6.0',
+    'cryptography>=2.8',
+    'distributed>=2.6.0',
+    'ConfigSpace<=0.4.10',
+    'graphviz<0.9.0,>=0.8.1',
```
Comment (mli) [inline core/setup.py:56]: how these `<` are decided, just blocking to use a future major release? if another library installed by the user requires a higher version, then there is a conflict.

File: docs/api/autogluon.core.space.rst [hunk 18]
```diff
@@ -11,7 +11,7 @@ Search space of possible hyperparameter values to consider.
    Define a dummy training function with searchable spaces for hyperparameters `lr` and `wd`:
 
    >>> import numpy as np
-   >>> import autogluon as ag
+   >>> import autogluon.core as ag
```
Comment (mli) [inline docs/api/autogluon.core.space.rst:14]: suggest to also test these codes. one option is using [doctest](https://docs.python.org/3/library/doctest.html). not necessary to do it right now, please consider enable it in a future PR if not enabled before.

File: text/tests/test_check_style.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
+from subprocess import Popen, PIPE
+import logging
+
+
+def test_check_style():
```
Comment (mli) [inline text/tests/test_check_style.py:5]: are we already enabled style check? it's very good to have it.

File: core/src/autogluon/core/loss.py (hunk 13 above)
Comment (Innixma) [reply core/src/autogluon/core/loss.py:5]: can do as follow-up PR to make this change more visible

File: .github/workflows/pypi_release.yml (hunk 8 above)
Comment (gradientsky) [reply .github/workflows/pypi_release.yml:59]: It's going to install all the packages in [requirements](https://github.com/gradientsky/autogluon/blob/modules-v2/autogluon/setup.py#L43-L73) including all the sub-modules.

File: Jenkinsfile (hunk 9 above)
Comment (gradientsky) [reply Jenkinsfile:110]: The build process builds packages in clean env in incremental order. If something is not in previous packages, the build will fail.

File: README.md (hunk 10 above)
Comment (gradientsky) [reply README.md:21]: Adding to the rest of the modules

File: autogluon/setup.py (hunk 11 above)
Comment (gradientsky) [reply autogluon/setup.py:44]: I double checked, it's there:
https://github.com/gradientsky/autogluon/blob/modules-v2/core/setup.py#L44
https://github.com/gradientsky/autogluon/blob/modules-v2/tabular/setup.py#L42
https://github.com/gradientsky/autogluon/blob/modules-v2/extra/setup.py#L44
https://github.com/gradientsky/autogluon/blob/modules-v2/mxnet/setup.py#L42
https://github.com/gradientsky/autogluon/blob/modules-v2/text/setup.py#L44
https://github.com/gradientsky/autogluon/blob/modules-v2/vision/setup.py#L44

File: docs/api/index.rst (hunk 15 above)
Comment (gradientsky) [reply docs/api/index.rst:11]: The docs work, though `autogluon.core.task` doesn't make sense because it's fractured into separate sub-modules: `autogluon.vision.task`, etc. I think we can refactor this in the follow-up PR. I'm currently trying to minimize any other changes beyond moving and necessary fixes to keep the code running.

File: docs/api/autogluon.core.space.rst (hunk 18 above)
Comment (gradientsky) [reply docs/api/autogluon.core.space.rst:14]: It is tested through d2lbook + sphinx. Big part of this CR was making sure that the docs compile.

File: text/tests/test_check_style.py (hunk 19 above)
Comment (gradientsky) [reply text/tests/test_check_style.py:5]: I would suggest not to add this in this PR because so many other things are changing.

File: core/setup.py (hunk 17 above)
Comment (Innixma) [reply core/setup.py:56]: unsure why the graphviz < was decided, but most tabular specific < are just to avoid new major releases that may have API breaking changes. We limit sklearn to minor version due to past experiences of it breaking API in minor version update.

File: Jenkinsfile [hunk 20]
```diff
@@ -16,9 +16,48 @@ stage("Unit Test") {
...
+        pip uninstall -y autogluon
+        pip uninstall -y autogluon.vision
+        pip uninstall -y autogluon.text
+        pip uninstall -y autogluon.mxnet
+        pip uninstall -y autogluon.extra
+        pip uninstall -y autogluon.tabular
+        pip uninstall -y autogluon.core
+
+        cd core/
```
Comment (Innixma) [inline Jenkinsfile:28]: one issue I had run into when benchmarking was I needed to upgrade setuptools to work with namespace packages.

```pip install -U setuptools```

This had to be done prior to any installs.

Perhaps we add this to install FAQ?

**Review by Innixma — COMMENTED** [review]
Benchmark results for Tabular show that everything is good, no errors or degradation in performance. I am fine to approve once we determine how we are going to surface the 0.0.14 documentation for users in the meantime to avoid user confusion.

The easiest (and hacky) way would be to link to the 0.0.14 docs in the mainline documentation, and potentially have it also link to 0.0.14 docs from our README -> http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-689/1/index.html

We could add a new option in the top right blue bar (next to API/Installation/Tutorials/GitHub) that says v0.0.14 documentation or Stable documentation, etc.

File: Jenkinsfile (hunk 20 above)
Comment (gradientsky) [reply Jenkinsfile:28]: added to readme and FAQ

File: README.md (hunk 10 above)
Comment (gradientsky) [reply README.md:21]: We should be good. Just added setuptools upgrade as Nick suggested.

File: core/src/autogluon/core/utils/utils.py (hunk 14 above)
Comment (gradientsky) [reply core/src/autogluon/core/utils/utils.py:9]: it's also used in grid_searcher (just ParameterGrid). However, the functions in `utils.py` are used across the core, so it's not easily extract-able.

File: Jenkinsfile (hunk 20 above)
Comment (gradientsky) [reply Jenkinsfile:28]: added link to the old docs 
image

**Comment by szha** [comment]
Job PR-694-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/8/index.html

**Comment by szha** [comment]
Job PR-694-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/9/index.html

**Comment by szha** [comment]
Job PR-694-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/10/index.html

**Comment by szha** [comment]
Job PR-694-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/13/index.html

**Comment by szha** [comment]
Job PR-694-14 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/14/index.html

**Comment by szha** [comment]
Job PR-694-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/15/index.html

**Comment by szha** [comment]
Job PR-694-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-694/16/index.html

**Review by jwmueller — APPROVED** [review]
